python/union_find.py

```diff
@@ -0,0 +1,126 @@
+from typing import List, Dict
+
+INITIAL_RANK = 0
+
+
+class Node:
+    """
+    Class implementing a node in an union-find data structure.
+    Stores the current node's rank and a reference to its parent node.
+    """
+
+    def __init__(self, parent_id: int, rank: int = INITIAL_RANK):
+        self._parent = parent_id
+        self._rank = rank
+
+    @property
+    def parent(self) -> int:
+        return self._parent
+
+    @parent.setter
+    def parent(self, x: int):
+        self._parent = x
+
+    @property
+    def rank(self) -> int:
+        return self._rank
+
+    @rank.setter
+    def rank(self, x: int):
+        self._rank = x
+
+
+class UnionFind:
+    """
+    Class implementing a disjoint-set data structure.
+    """
+
+    def __init__(self, node_ids: List[int]):
+        self.nodes: Dict[int, Node] = {node_id: Node(parent_id=node_id, rank=INITIAL_RANK) for node_id in node_ids}
+
+    def parent(self, node_id: int) -> int:
+        """
+        Returns given node's parent's ID.
+
+        :param node_id: Node ID
+        """
+        return self.nodes[node_id].parent
+
+    def grandparent(self, node_id: int) -> int:
+        """
+        Returns given node's grandparent's ID.
+
+        :param node_id: Node ID
+        """
+        return self.parent(self.parent(node_id))
+
+    def rank(self, node_id: int) -> int:
+        """
+        Returns given node's rank.
+
+        :param node_id: Node ID
+        """
+        return self.nodes[node_id].rank
+
+    def find(self, node_id: int) -> int:
+        """
+        Returns the representative node's ID for the component that given node is member of.
+        Uses path splitting (https://en.wikipedia.org/wiki/Disjoint-set_data_structure#Finding_set_representatives)
+        in order to keep trees representing connected components flat.
+
+        :param node_id: Node ID
+        """
+        while node_id != self.parent(node_id):
+            self.nodes[node_id].parent = self.grandparent(node_id)  # path splitting
+            node_id = self.parent(node_id)
```

Comment:
Having hard time to understand this block. We are iterating as long as `node_id` is different than `node_id`'s parent, and at the line `#75` we are setting it to that exact value?

Reply:
This code moves down the tree towards its root.

Suppose that `node_id` is initially 9 in [this tree](github url comment). Its parent is 5 and thus `node_id` is then set to 5.
In the next iteration, `node_id` is 5 and `self.parent(node_id)` returns 2. The traversal continues until `node_id` reaches the root, whose parent is by convention identical to itself.

(For the sake of brevity, this explanation disregards line 74 but it applies in that case as well.)